test/controllers/users_controller_test.rb

```diff
@@ -1,4 +1,101 @@
 require "test_helper"
 
 describe UsersController do
+  describe "guest user (not authenticated)" do
+    describe "index" do
+      it "should display all users" do
+        get users_path
+        
+        must_respond_with :success
+      end
+      
+      it "should not break if there are no users" do 
+        User.destroy_all
+        
+        get users_path
+        
+        must_respond_with :success
+      end
+    end
+    
+    describe "show" do
+      it "should respond with success when given a valid user" do
+        valid_user = users(:dan)
+        
+        get user_path(valid_user.id)
+        must_respond_with :success
+      end
+      
+      it "should respond with redirect with an invalid user" do 
+        get user_path(-1)
+        
+        must_respond_with :not_found
+      end
+    end
+  end
+  
+  describe "authenticated user" do
+    before do
+      perform_login
+    end
+    
+    describe "logout" do
+      it "successfully logs out current user" do
+        expect(session[:user_id].nil?).must_equal false
+        
+        delete logout_path
+        
+        assert_nil(session[:user_id])
+        expect(flash[:success]).must_equal "Successfully logged out!"
+        
+        must_respond_with :redirect
+        must_redirect_to root_path
+      end
```

Comment:
according to the rubric we also need a test for performing logout when not logged in